### PR TITLE
#121 add ic regimen mappings

### LINKED ISSUE #121: Add support for Ivory Coast HIV indicators
See related desktop issues:

- https://github.com/sussol/msupply/pull/5389.
- https://github.com/sussol/msupply/issues/5410.

## Summary

Add HIV indicators to eSIGL service.

## Spec 

Current spec for IC HIV indicators includes the following columns:

```
[
  patients_adults_reçus,
  patients_enfants_reçus,
  patient_stable_adulte,
  patient_stable_enfant,
  nouvelle_inclusion_adulte,
  nouvelle_inclusion_enfante,
  referes
]
```

## eSIGL (VM)

Querying the current eSIGL VM for `PNLS-ANTIRETROVIRAUX ET IO`:

```
GET https://83.96.240.209/programId/25/programRegimenTemplate.json
```

```
{
    "template": {
        "programId": 25,
        "columns": [
            {
                "name": "patientsOnTreatmentAdult",
                "label": "Nombre de patients ADULTES reçus",
                "visible": true,
            },
            {
                "name": "patientsOnTreatmentChildren",
                "label": "Nombre de patients ENFANTS reçus",
                "visible": true,
            },
            {
                "name": "patientsToInitiateTreatmentAdult",
                "label": "Nombre de patients ADULTES naifs  (Inclusions)",
                "visible": true,
            },
            {
                "name": "patientsToInitiateTreatmentChildren",
                "label": "Nombre de patients ENFANTS naifs (Inclusions)",
                "visible": true,
            },
            {
                "name": "remarks",
                "label": "Remarques",
                "visible": true,
            },
            {
                "name": "patientsOnTreatment",
                "label": "Number of patients on treatment",
                "visible": false,
            },
            {
                "name": "patientsToInitiateTreatment",
                "label": "Number of patients to be initiated treatment",
                "visible": false,
            },
            {
                "name": "patientsStoppedTreatment",
                "label": "Number of patients stopped treatment",
                "visible": false,
            },
            {
                "name": "patientsStoppedTreatmentAdult",
                "label": "Patients Stopped Treatment Adult",
                "visible": false,
            },
            {
                "name": "patientsStoppedTreatmentChildren",
                "label": "Patients Stopped Treatment Children",
                "visible": false,
            }
        ]
    }
}
```

Partial column matching (`indicator_attribute.code` <-> `column.name`):

```
patients_adults_reçus <-> patientsOnTreatmentAdult
patients_enfants_reçus <-> patientsOnTreatmentChildren
patient_stable_adulte <-> ???
patient_stable_enfant <-> ???
nouvelle_inclusion_adulte <-> patientsToInitiateTreatmentAdult
nouvelle_inclusion_enfante <-> patientsToInitiateTreatmentChildren
referes <-> remarks
```

Unmatched columns (HIV spec):

```
patient_stable_adulte
patient_stable_enfant
```

Unmatched columns (eSIGL):

```
patientsOnTreatment
patientsToInitiateTreatment
patientsStoppedTreatment
patientsStoppedTreatmentAdult
patientsStoppedTreatmentChildren
```

## eSIGL (current)

From Kevin (06/02):

IC-HIV-eSIGL-kevin

Column matching (`indicator_attribute.code` <-> `column.label`):

```
patients_adults_reçus <-> "Nombre de patients ADULTES reçus"
patients_enfants_reçus <-> "Nombre de patients ENFANTS reçus"
patient_stable_adulte <-> "Nombre de patients stables Adulte (>= 15 ans) reçus"
patient_stable_enfant <->  "Nombre de patients stables Enfant (< 15 ans) reçus"
nouvelle_inclusion_adulte <-> "Nombe de patients ADULTES naifs (Inclusions)"
nouvelle_inclusion_enfante <-> "Nombe de patients ENFANTS naifs (Inclusions)"
referes <-> "Référés (Fiche de référence et contre référence/ Dossiers patients)"
```

## Implementation

- update format of incoming requisition (desktop issue: https://github.com/sussol/msupply/pull/5389).
- update `requisitionMerge` to handle columns.

### PR DESCRIPTION
**Pull request by wlthomson**
Fixes #121.

Provisional PR based on existing information. Some column mappings will likely need to be updated before can be put into production.

### CODE REVIEW

File: src/requisitionMerge.js [hunk 1]
```diff
@@ -20,6 +20,67 @@ const MERGE_FIELDS_MAPPING = {
...
+        ({ indicator_ID: rowIndicatorID }) => rowIndicatorID === indicator.ID
+      );
+      const indicatorColumns = columns.filter(
+        ({ indicator_ID: columnIndicatorID }) => columnIndicatorID === indicator.ID
+      );
+      // Convert each indicator row to matching eSIGL regimen line.
+      const indicatorRegimenLineItems = indicatorRows.map(indicatorRow => {
+        const { code } = indicatorRow;
+        const regimenColumns = indicatorColumns.map(indicatorColumn => {
+          const [{ value: columnValue }] = values.filter(
```
Comment (josh-griffin) [inline src/requisitionMerge.js]: why not `.find`?

File: src/requisitionMerge.js [hunk 2]
```diff
@@ -163,22 +224,33 @@ function requisitionItemsMerge(incomingRequisitionLines, outgoingRequisitionLine
...
-  if (regimenLineItems) {
-    regimenLineItems.forEach(regimenItem => {
-      const matchingItem = regimenData.find(regimenDatum => regimenDatum.code === regimenItem.code);
-      const value = !matchingItem ? 0 : matchingItem.value || 0;
-      regimenItem.patientsOnTreatment = value;
-    });
-  }
+  const regimenLineItems = outgoingRegimenLineItems.map(outgoingRegimenLineItem => {
+    // Populate each outgoing regimen line with incoming regimen line value. If no matching
+    // incoming regimen line exists, use default value.
```
Comment (josh-griffin) [inline src/requisitionMerge.js]: just a comment here that the item lines with no match are being sent back as unmatched https://github.com/openmsupply/elmis-service/blob/8743b102bdc3b150153d9071996a6efef6a30cf8/src/requisitionMerge.js#L216 - I think this is a small change and can offer huge aid when coming to debug

**Review by josh-griffin — APPROVED** [review]
Haven't tested but 🎉 

Let me know if you want me to give it a whirl?

**Comment by mark-prins** [comment]
I agree with @joshxg - haven't tested, but lets give it a go!
Looks great to me :)

File: src/requisitionMerge.js (hunk 2 above)
Comment (wlthomson) [reply src/requisitionMerge.js]: e59eea6f2b682891e0db39a8f6fb19c310bd03cd. 

Not the nicest yet, but when time permits can make it a bit more robust. Problem is that need to be able to track mismatched regimens as well as mismatched columns (currently only detects mismatched regimens).

File: src/requisitionMerge.js (hunk 1 above)
Comment (wlthomson) [reply src/requisitionMerge.js]: 28be16180f14adbccdd47500f34994e121417a0f.

File: src/tests/fixtures/regimens.js [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+const REGIMEN_COLUMNS = {
+  VALUE: 'value',
+  COMMENT: 'comment',
+};
+
+const HIV_COLUMNS_MSUPPLY = {
+  [HIV_COLUMNS.PATIENT_ADMITTED_ADULT]: 'patients_adultes_recus',
+  [HIV_COLUMNS.PATIENT_ADMITTED_CHILD]: 'patients_enfants_recus',
+  // [HIV_COLUMNS.PATIENT_STABLE_ADULT]: 'patient_stable_adulte',
+  // [HIV_COLUMNS.PATIENT_STABLE_CHILD]: 'patients_stable_enfant',
```
Comment (josh-griffin) [inline src/tests/fixtures/regimens.js:24]: oops?

File: src/tests/fixtures/regimens.js [hunk 4]
```diff
@@ -0,0 +1,201 @@
...
+const REGIMEN_COLUMNS_MSUPPLY = {
+  [REGIMEN_COLUMNS.VALUE]: 'regimen_value',
+  [REGIMEN_COLUMNS.COMMENT]: 'regimen_comment',
+};
+
+const HIV_COLUMNS_ESIGL = {
+  [HIV_COLUMNS.PATIENT_ADMITTED_ADULT]: 'patientsOnTreatmentAdult',
+  [HIV_COLUMNS.PATIENT_ADMITTED_CHILD]: 'patientsOnTreatmentChildren',
+  // [HIV_COLUMNS.PATIENT_STABLE_ADULT]: '',
+  // [HIV_COLUMNS.PATIENT_STABLE_CHILD]: '',
```
Comment (josh-griffin) [inline src/tests/fixtures/regimens.js:39]: oops

File: src/requisitionMerge.js [hunk 5]
```diff
@@ -90,6 +160,39 @@ const findMatchedRequisition = ({ code: incomingItemCode }) => ({
...
 
+/**
+ * Merges an array of incoming regimen lines with outgoing regimen lines.
+ *
+ * @param  {Array} incomingRegimenLines Regimen lines of the incoming requisition
+ * @param  {Array} outgoingRegimenLines Regimen lines of the outgoing requisition
+ * @return {Array} The merged regimen lines.
+ */
+const requisitionRegimensMerge = (incomingRegimenLines, outgoingRegimenLines) => {
+  const incomingLines = [...incomingRegimenLines.map(getMappedRegimenColumns)];
```
Comment (josh-griffin) [inline src/requisitionMerge.js]: I don't think you need this spread

File: src/requisitionMerge.js [hunk 6]
```diff
@@ -90,6 +160,39 @@ const findMatchedRequisition = ({ code: incomingItemCode }) => ({
...
+  const outgoingLines = [...outgoingRegimenLines];
+
+  const outgoingRegimenCodes = new Set(outgoingLines.map(({ code }) => code));
+  const matchingRegimenCodes = new Set(
```
Comment (josh-griffin) [inline src/requisitionMerge.js]: isn't this a set of objects, not just codes

**Review by josh-griffin — CHANGES_REQUESTED** [review]
hmm am I crazy, how is this passing 😕

File: src/tests/fixtures/regimens.js (hunk 3 above)
Comment (wlthomson) [reply src/tests/fixtures/regimens.js:24]: Intended, just a placeholder for now.

File: src/requisitionMerge.js (hunk 5 above)
Comment (wlthomson) [reply src/requisitionMerge.js]: 8f8fef9e952881b28a06d3053f87028697c67b43.

File: src/requisitionMerge.js (hunk 6 above)
Comment (wlthomson) [reply src/requisitionMerge.js]: 8f8fef9e952881b28a06d3053f87028697c67b43.

**Comment by wlthomson** [comment]
> hmm am I crazy, how is this passing 😕

Have you given it a test? All should be passing now 😄 .

On a side note, would be a nice idea to integrate our tests with pre-commit or push hooks (or setup up CI, but that's a bit intense for this size project).

**Review by josh-griffin — APPROVED** [review]
LGTM

**Pull request merged**